src/k8s_upgrade.py

```diff
@@ -0,0 +1,497 @@
+        )
+        self.run_post_upgrade_checks(event, finished_whole_cluster=False)
+
+        if self._upgrade.unit_state != UnitState.HEALTHY:
+            logger.info(
+                f"Unit state is not healhy but {self._upgrade.unit_state}, not continuing post-refresh checks."
+            )
+            return
+
+        logger.debug("Cluster is healthy after refreshing unit %s", self.charm.unit.name)
+        self.charm.status.set_and_share_status(ActiveStatus())
+
+        # Leader of config-server must wait for all shards to be upgraded before finalising the
+        # upgrade.
+        if not self.charm.unit.is_leader() or not self.charm.is_role(Config.Role.CONFIG_SERVER):
+            logger.debug("Post refresh check is completed.")
+            return
+
+        self.charm.upgrade.post_cluster_upgrade_event.emit()
+
+    def run_post_cluster_upgrade_task(self, event: EventBase) -> None:
+        """Waits for entire cluster to be upgraded before enabling the balancer."""
+        # Leader of config-server must wait for all shards to be upgraded before finalising the
+        # upgrade.
+        if not self.charm.unit.is_leader() or not self.charm.is_role(Config.Role.CONFIG_SERVER):
+            return
+
+        if not self.charm.is_cluster_on_same_revision():
+            logger.debug("Waiting to finalise refresh, one or more shards need refresh.")
+            event.defer()
+            return
+
+        logger.debug(
+            "Entire cluster has been refreshed, checking health of the cluster and enabling balancer."
+        )
+        self.run_post_upgrade_checks(event, finished_whole_cluster=True)
+
+        try:
+            with MongosConnection(self.charm.mongos_config) as mongos:
+                mongos.start_and_wait_for_balancer()
+        except BalancerNotEnabledError:
+            logger.debug(
+                "Need more time to enable the balancer after finishing the refresh. Deferring event."
+            )
+            event.defer()
+            return
+
+        self.set_mongos_feature_compatibilty_version(Config.Upgrade.FEATURE_VERSION_6)
+
+    def _on_pre_upgrade_check_action(self, event: ActionEvent) -> None:
+        """Runs the pre-refresh checks to ensure that the deployment is ready for refresh."""
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader {PRECHECK_ACTION_NAME}`)"
+            logger.debug(f"Pre-refresh check failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or self._upgrade.in_progress:
+            message = "Upgrade already in progress"
+            logger.debug(f"Pre-refresh check failed: {message}")
+            event.fail(message)
+            return
+        try:
+            self._upgrade.pre_upgrade_check()
+            self.charm.status.process_statuses()
```

Comment:
noobie question, why is this needed in K8s but not VM?

Reply:
I tried something that did not work as expected and I forgot to remove it :)
Removing it